blocks/certifier.go

```diff
@@ -0,0 +1,413 @@
+		logger.Error("block not registered for cert")
+		return unknown, errInternal
+	}
+
+	if time.Now().After(c.certMsgs[lid].deadline) {
+		logger.With().Debug("cert msg received after deadline", log.Time("deadline", c.certMsgs[lid].deadline))
+		return expired, nil
+	}
+	if c.certMsgs[lid].done {
+		return certified, nil
+	}
+	return pending, nil
+}
+
+// HandleSyncedCertificate handles Certificate from sync.
+func (c *Certifier) HandleSyncedCertificate(ctx context.Context, lid types.LayerID, cert *types.Certificate) error {
+	logger := c.logger.WithContext(ctx).WithFields(cert.BlockID)
+	for _, msg := range cert.Signatures {
+		if err := c.validate(ctx, logger, msg); err != nil {
+			return err
+		}
+	}
+	if err := c.save(logger, lid, cert); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Certifier) handleRawCertifyMsg(ctx context.Context, data []byte) error {
+	logger := c.logger.WithContext(ctx)
+	var msg types.CertifyMessage
+	if err := codec.Decode(data, &msg); err != nil {
+		logger.With().Error("malformed cert msg", log.Err(err))
+		return errMalformedData
+	}
+	if err := c.validate(ctx, logger, msg); err != nil {
+		return err
+	}
+	if err := c.saveMessage(logger, msg); err != nil {
+		return errInternal
+	}
+	return nil
+}
+
+func (c *Certifier) validate(ctx context.Context, logger log.Log, msg types.CertifyMessage) error {
+	lid := msg.LayerID
+	bid := msg.BlockID
+	if state, err := c.getCertState(logger, lid, bid); err != nil {
+		return err
+	} else if state == expired {
+		return errMsgTooOld
+	} else if state == certified {
+		// should still gossip this msg to peers even when this node has created a certificate
+		return nil
+	}
+
+	logger.WithFields(bid)
+	// extract public key from signature
+	pubkey, err := ed25519.ExtractPublicKey(msg.Bytes(), msg.Signature)
+	if err != nil {
+		return fmt.Errorf("%w: cert msg extract key: %v", errMalformedData, err.Error())
+	}
+	nid := types.BytesToNodeID(pubkey)
+	valid, err := c.oracle.Validate(ctx, msg.LayerID, eligibility.CertifyRound, c.cfg.CommitteeSize, nid, msg.Proof, msg.EligibilityCnt)
+	if err != nil {
+		logger.With().Warning("failed to validate cert msg", log.Err(err))
+		return errInternal
+	}
+	if !valid {
+		logger.With().Warning("invalid cert msg", log.Err(err))
+		return errInvalidCertMsg
+	}
+	return nil
+}
+
+func (c *Certifier) saveMessage(logger log.Log, msg types.CertifyMessage) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	lid := msg.LayerID
+	bid := msg.BlockID
+	if state, err := c.getCertStateLocked(logger, lid, bid); err != nil {
+		return err
+	} else if state != pending { // already certified
+		return nil
+	}
+
+	if _, ok := c.certMsgs[lid]; !ok {
+		logger.Fatal("missing layer in cache")
+	}
+	c.certMsgs[lid].signatures = append(c.certMsgs[lid].signatures, msg)
+
+	if created, err := c.tryGenCertLocked(logger, lid, bid, c.certMsgs[lid].signatures); err != nil {
+		return err
+	} else if created {
+		c.certMsgs[lid].done = true
+	}
+	return nil
+}
+
```

Comment:
i don't understand the purpose of this method
certificate should be generatable when handler received sufficient number of messages to cross threshold